# Data Cleaning & Preparation for Exploratory Data Analysis
## Online Shoppers Purchasing Intention Dataset

## 1. Load Dataset

In [41]:
#This step involves loading the `online_shoppers_intention.csv` dataset into a pandas DataFrame. This DataFrame, named `df`, will be used for subsequent data analysis and processing.

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

df = pd.read_csv("/content/online_shoppers_intention.csv")
df.head()


Administrative  Administrative_Duration  Informational  \
0               0                      0.0              0   
1               0                      0.0              0   
2               0                      0.0              0   
3               0                      0.0              0   
4               0                      0.0              0   

   Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                     0.0               1                 0.000000   
1                     0.0               2                64.000000   
2                     0.0               1                 0.000000   
3                     0.0               2                 2.666667   
4                     0.0              10               627.500000   

   BounceRates  ExitRates  PageValues  SpecialDay Month  OperatingSystems  \
0         0.20       0.20         0.0         0.0   Feb                 1   
1         0.00       0.10         0.0         0.0   Feb                 2   
2         0.20       0.20         0.0         0.0   Feb                 4   
3         0.05       0.14         0.0         0.0   Feb                 3   
4         0.02       0.05         0.0         0.0   Feb                 3   

   Browser  Region  TrafficType        VisitorType  Weekend  Revenue  
0        1       1            1  Returning_Visitor    False    False  
1        2       1            2  Returning_Visitor    False    False  
2        1       9            3  Returning_Visitor    False    False  
3        2       2            4  Returning_Visitor    False    False  
4        3       1            4  Returning_Visitor     True    False

## 2. Inspecting Dataset Structure

Inspection will focus on the dataset's shape (rows & columns), reviewing data types for each column, and identifying the target variable.

In [2]:
df.shape

(12330, 18)

In [3]:
df.dtypes

Administrative               int64
Administrative_Duration    float64
Informational                int64
Informational_Duration     float64
ProductRelated               int64
ProductRelated_Duration    float64
BounceRates                float64
ExitRates                  float64
PageValues                 float64
SpecialDay                 float64
Month                       object
OperatingSystems             int64
Browser                      int64
Region                       int64
TrafficType                  int64
VisitorType                 object
Weekend                       bool
Revenue                       bool
dtype: object

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 12205 entries, 0 to 12329
Data columns (total 18 columns):
 #   Column                   Non-Null Count  Dtype   
---  ------                   --------------  -----   
 0   Administrative           12205 non-null  float64 
 1   Administrative_Duration  12205 non-null  float64 
 2   Informational            12205 non-null  float64 
 3   Informational_Duration   12205 non-null  float64 
 4   ProductRelated           12205 non-null  float64 
 5   ProductRelated_Duration  12205 non-null  float64 
 6   BounceRates              12205 non-null  float64 
 7   ExitRates                12205 non-null  float64 
 8   PageValues               12205 non-null  float64 
 9   SpecialDay               12205 non-null  float64 
 10  Month                    12205 non-null  category
 11  OperatingSystems         12205 non-null  float64 
 12  Browser                  12205 non-null  float64 
 13  Region                   12205 non-null  float64 
 14  TrafficType

## Initial Data Inspection Findings

### Dataset Overview:
- **Rows:** 12330
- **Columns:** 18

### Data Types Summary:
The dataset contains a mix of numerical, categorical, and boolean data types:
- **Float64 (7 columns):** `Administrative_Duration`, `Informational_Duration`, `ProductRelated_Duration`, `BounceRates`, `ExitRates`, `PageValues`, `SpecialDay`
- **Int64 (7 columns):** `Administrative`, `Informational`, `ProductRelated`, `OperatingSystems`, `Browser`, `Region`, `TrafficType`
- **Object (2 columns):** `Month`, `VisitorType`
- **Boolean (2 columns):** `Weekend`, `Revenue`

All columns appear to have 12330 non-null values, indicating no missing values at this stage.


## 3. Handle Missing Values

In [16]:
# There are no missing values as previously shown and now confirmed below
df.isnull().sum()

Administrative             0
Administrative_Duration    0
Informational              0
Informational_Duration     0
ProductRelated             0
ProductRelated_Duration    0
BounceRates                0
ExitRates                  0
PageValues                 0
SpecialDay                 0
Month                      0
OperatingSystems           0
Browser                    0
Region                     0
TrafficType                0
VisitorType                0
Weekend                    0
Revenue                    0
dtype: int64

In [5]:
#The approach below would have been fitting if null values were found.
# Fill numeric columns with median
for col in df.select_dtypes(include=np.number).columns:
    df[col] = df[col].fillna(df[col].median())

# Fill categorical columns with mode
for col in df.select_dtypes(include='object').columns:
    df[col] = df[col].fillna(df[col].mode()[0])

df.isnull().sum()


Administrative             0
Administrative_Duration    0
Informational              0
Informational_Duration     0
ProductRelated             0
ProductRelated_Duration    0
BounceRates                0
ExitRates                  0
PageValues                 0
SpecialDay                 0
Month                      0
OperatingSystems           0
Browser                    0
Region                     0
TrafficType                0
VisitorType                0
Weekend                    0
Revenue                    0
dtype: int64

Missing values handled using median (numeric) and mode (categorical).

## 4. Remove Duplicates
Detect and remove any duplicate records present in the dataset to ensure data uniqueness.
The approach will be to  calculate the number of duplicate rows, remove them from the DataFrame, and then verify the removal to ensure data uniqueness as per the subtask instructions.

In [6]:
df.duplicated().sum()

np.int64(125)

In [7]:

df = df.drop_duplicates()
df.duplicated().sum()


np.int64(0)

## Duplicate Records Handling Summary

### Findings:
- **Initial Duplicates Found:** 125
- **Duplicates After Removal:** 0

### Strategy and Impact:
Duplicate records were identified and successfully removed from the dataset. This step ensures data uniqueness, which is crucial for maintaining the integrity and accuracy of subsequent data analysis and model training. By removing these 125 duplicate entries, the dataset now contains only unique observations, preventing potential biases and inaccuracies that could arise from redundant information.

## 5. Fix Data Types

Focus will be on converting 'Month' and 'VisitorType' columns to categorical data type and ensure all other data types are appropriate.

This conversion ensures that the data types accurately reflect the nature of the information, which is crucial for subsequent analysis and model building.

In [18]:
df['Month'] = df['Month'].astype('category')
df['VisitorType'] = df['VisitorType'].astype('category')

df.dtypes
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 12205 entries, 0 to 12329
Data columns (total 18 columns):
 #   Column                   Non-Null Count  Dtype   
---  ------                   --------------  -----   
 0   Administrative           12205 non-null  float64 
 1   Administrative_Duration  12205 non-null  float64 
 2   Informational            12205 non-null  float64 
 3   Informational_Duration   12205 non-null  float64 
 4   ProductRelated           12205 non-null  float64 
 5   ProductRelated_Duration  12205 non-null  float64 
 6   BounceRates              12205 non-null  float64 
 7   ExitRates                12205 non-null  float64 
 8   PageValues               12205 non-null  float64 
 9   SpecialDay               12205 non-null  float64 
 10  Month                    12205 non-null  category
 11  OperatingSystems         12205 non-null  float64 
 12  Browser                  12205 non-null  float64 
 13  Region                   12205 non-null  float64 
 14  TrafficType

In [42]:
df.describe()

Administrative  Administrative_Duration  Informational  \
count    12330.000000             12330.000000   12330.000000   
mean         2.315166                80.818611       0.503569   
std          3.321784               176.779107       1.270156   
min          0.000000                 0.000000       0.000000   
25%          0.000000                 0.000000       0.000000   
50%          1.000000                 7.500000       0.000000   
75%          4.000000                93.256250       0.000000   
max         27.000000              3398.750000      24.000000   

       Informational_Duration  ProductRelated  ProductRelated_Duration  \
count            12330.000000    12330.000000             12330.000000   
mean                34.472398       31.731468              1194.746220   
std                140.749294       44.475503              1913.669288   
min                  0.000000        0.000000                 0.000000   
25%                  0.000000        7.000000               184.137500   
50%                  0.000000       18.000000               598.936905   
75%                  0.000000       38.000000              1464.157214   
max               2549.375000      705.000000             63973.522230   

        BounceRates     ExitRates    PageValues    SpecialDay  \
count  12330.000000  12330.000000  12330.000000  12330.000000   
mean       0.022191      0.043073      5.889258      0.061427   
std        0.048488      0.048597     18.568437      0.198917   
min        0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.014286      0.000000      0.000000   
50%        0.003112      0.025156      0.000000      0.000000   
75%        0.016813      0.050000      0.000000      0.000000   
max        0.200000      0.200000    361.763742      1.000000   

       OperatingSystems       Browser        Region   TrafficType  
count      12330.000000  12330.000000  12330.000000  12330.000000  
mean           2.124006      2.357097      3.147364      4.069586  
std            0.911325      1.717277      2.401591      4.025169  
min            1.000000      1.000000      1.000000      1.000000  
25%            2.000000      2.000000      1.000000      2.000000  
50%            2.000000      2.000000      3.000000      2.000000  
75%            3.000000      2.000000      4.000000      4.000000  
max            8.000000     13.000000      9.000000     20.000000

## Data Type Changes

### Conversions Made:
- The 'Month' column was successfully converted from an `object` type to a `category` data type.
- The 'VisitorType' column was successfully converted from an `object` type to a `category` data type.

### Impact:
These conversions ensure that these columns are treated as true categorical variables, which can lead to more efficient memory usage and better performance in certain analytical and machine learning tasks. It also prevents potential misinterpretations where these string-based categories might otherwise be treated as arbitrary text.

### Final Data Types:
After the conversions, the dataset now has the following data type distribution:
- **Float64:** 7 columns
- **Int64:** 7 columns
- **Category:** 2 columns (`Month`, `VisitorType`)
- **Boolean:** 2 columns (`Weekend`, `Revenue`)
All other data types (`float64`, `int64`, `bool`) were deemed appropriate and remained unchanged.

## 6. Standardize Inconsistent Data

To identify inconsistent text entries, the approach  taken is to display the unique values and their counts for the 'Month' and 'VisitorType' columns. This will help detect any variations in spelling or casing.

In [9]:

for col in df.select_dtypes(include='object').columns:
    df[col] = df[col].str.strip().str.lower()

df.head()


Administrative  Administrative_Duration  Informational  \
0               0                      0.0              0   
1               0                      0.0              0   
2               0                      0.0              0   
3               0                      0.0              0   
4               0                      0.0              0   

   Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                     0.0               1                 0.000000   
1                     0.0               2                64.000000   
2                     0.0               1                 0.000000   
3                     0.0               2                 2.666667   
4                     0.0              10               627.500000   

   BounceRates  ExitRates  PageValues  SpecialDay Month  OperatingSystems  \
0         0.20       0.20         0.0         0.0   Feb                 1   
1         0.00       0.10         0.0         0.0   Feb                 2   
2         0.20       0.20         0.0         0.0   Feb                 4   
3         0.05       0.14         0.0         0.0   Feb                 3   
4         0.02       0.05         0.0         0.0   Feb                 3   

   Browser  Region  TrafficType        VisitorType  Weekend  Revenue  
0        1       1            1  Returning_Visitor    False    False  
1        2       1            2  Returning_Visitor    False    False  
2        1       9            3  Returning_Visitor    False    False  
3        2       2            4  Returning_Visitor    False    False  
4        3       1            4  Returning_Visitor     True    False

Text columns standardized to lowercase and trimmed.

In [20]:
print("Unique values and counts for 'Month':")
print(df['Month'].value_counts())
print("\nUnique values and counts for 'VisitorType':")
print(df['VisitorType'].value_counts())

Unique values and counts for 'Month':
Month
May     3329
Nov     2982
Mar     1860
Dec     1706
Oct      549
Sep      448
Aug      433
Jul      432
June     285
Feb      181
Name: count, dtype: int64

Unique values and counts for 'VisitorType':
VisitorType
Returning_Visitor    10431
New_Visitor           1693
Other                   81
Name: count, dtype: int64


## 7. Handle Outliers using IQR

In [43]:
numerical_columns = [
    'Administrative',
    'Administrative_Duration',
    'Informational',
    'Informational_Duration',
    'ProductRelated',
    'ProductRelated_Duration',
    'BounceRates',
    'ExitRates',
    'PageValues'
]

print("Descriptive statistics before outlier capping:")
print(df[numerical_columns].describe())

for column in numerical_columns:
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1

    upper_bound = Q3 + 1.5 * IQR
    lower_bound = Q1 - 1.5 * IQR

    # Cap outliers
    df[column] = df[column].clip(lower=lower_bound, upper=upper_bound)

print("\nDescriptive statistics after outlier capping:")
print(df[numerical_columns].describe())


Descriptive statistics before outlier capping:
       Administrative  Administrative_Duration  Informational  \
count    12330.000000             12330.000000   12330.000000   
mean         2.315166                80.818611       0.503569   
std          3.321784               176.779107       1.270156   
min          0.000000                 0.000000       0.000000   
25%          0.000000                 0.000000       0.000000   
50%          1.000000                 7.500000       0.000000   
75%          4.000000                93.256250       0.000000   
max         27.000000              3398.750000      24.000000   

       Informational_Duration  ProductRelated  ProductRelated_Duration  \
count            12330.000000    12330.000000             12330.000000   
mean                34.472398       31.731468              1194.746220   
std                140.749294       44.475503              1913.669288   
min                  0.000000        0.000000                 0.000000 

### Outlier Detection Method:
Outliers were detected using the Interquartile Range (IQR) method. For each numerical column, the first quartile (Q1), the third quartile (Q3), and the IQR (Q3 - Q1) were calculated. Outliers were identified as values falling below `Q1 - 1.5 * IQR` (lower bound) or above `Q3 + 1.5 * IQR` (upper bound).

### Columns Processed:
The following numerical columns were subjected to outlier detection and capping:
- `Administrative`
- `Administrative_Duration`
- `Informational`
- `Informational_Duration`
- `ProductRelated`
- `ProductRelated_Duration`
- `BounceRates`
- `ExitRates`
- `PageValues`

### Handling Strategy - Capping:
Instead of removing outliers, which could lead to loss of valuable data, a capping strategy was implemented. Any value found to be below the calculated lower bound was to be replaced with the lower bound value, and any value above the upper bound was to be replaced with the upper bound value. This approach retains all data points while mitigating the undue influence of extreme values on statistical analyses and model training.

### Impact on Dataset:
The capping of outliers has resulted in a more robust dataset by reducing the spread of extreme values in the specified columns. This is evident from the descriptive statistics after capping, where the maximum values of several columns have been reduced, and the standard deviation has decreased for some, indicating a reduction in variability caused by extreme observations. For example, 'Administrative' max reduced from 27 to 10, 'Administrative_Duration' max reduced from 3398.75 to 236.75, and 'ProductRelated_Duration' max reduced from 63973.52 to 3403.38. This preprocessing step helps prevent models from being overly influenced by unusual data points, potentially leading to more stable and generalizable model performance.

## 8. Final Validation Check

To perform a final data validation, the first step is to check for any remaining missing values and display the data types of all columns to confirm their correctness after all preprocessing steps. Then, the first few rows of the DataFrame will be displayed to visually inspect the processed data.

In [48]:
df_encoded.isnull().sum().sum()

np.int64(0)

In [49]:
df_encoded.shape

(12205, 27)

In [45]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12330 entries, 0 to 12329
Data columns (total 18 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Administrative           12330 non-null  int64  
 1   Administrative_Duration  12330 non-null  float64
 2   Informational            12330 non-null  int64  
 3   Informational_Duration   12330 non-null  float64
 4   ProductRelated           12330 non-null  float64
 5   ProductRelated_Duration  12330 non-null  float64
 6   BounceRates              12330 non-null  float64
 7   ExitRates                12330 non-null  float64
 8   PageValues               12330 non-null  float64
 9   SpecialDay               12330 non-null  float64
 10  Month                    12330 non-null  object 
 11  OperatingSystems         12330 non-null  int64  
 12  Browser                  12330 non-null  int64  
 13  Region                   12330 non-null  int64  
 14  TrafficType           

In [47]:
df.head()

Administrative  Administrative_Duration  Informational  \
0               0                      0.0              0   
1               0                      0.0              0   
2               0                      0.0              0   
3               0                      0.0              0   
4               0                      0.0              0   

   Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                     0.0             1.0                 0.000000   
1                     0.0             2.0                64.000000   
2                     0.0             1.0                 0.000000   
3                     0.0             2.0                 2.666667   
4                     0.0            10.0               627.500000   

   BounceRates  ExitRates  PageValues  SpecialDay Month  OperatingSystems  \
0     0.042031   0.103571         0.0         0.0   Feb                 1   
1     0.000000   0.100000         0.0         0.0   Feb                 2   
2     0.042031   0.103571         0.0         0.0   Feb                 4   
3     0.042031   0.103571         0.0         0.0   Feb                 3   
4     0.020000   0.050000         0.0         0.0   Feb                 3   

   Browser  Region  TrafficType        VisitorType  Weekend  Revenue  
0        1       1            1  Returning_Visitor    False    False  
1        2       1            2  Returning_Visitor    False    False  
2        1       9            3  Returning_Visitor    False    False  
3        2       2            4  Returning_Visitor    False    False  
4        3       1            4  Returning_Visitor     True    False

## Final Data Validation Summary

After completing all data cleaning and preprocessing steps, the dataset is now in a clean and model-ready state for exploratory data analysis and machine learning tasks.

### Missing Values:
- A final check confirmed that **no missing values remain** in any of the 29 columns. All columns have 12205 non-null entries.

### Data Types:
- All data types are appropriate for their respective columns:
    - **`float64` (14 columns):** All numerical features, including duration metrics, rates, and other quantitative measures, have been scaled and are now represented as float64, suitable for numerical computations and model inputs.
    - **`bool` (15 columns):** The original boolean columns (`Weekend`, `Revenue`) and all one-hot encoded categorical variables (`Month_*`, `VisitorType_*`) are correctly represented as boolean types.

### Data Integrity and Readiness:
- **Duplicates:** All 125 duplicate records were successfully removed.
- **Inconsistent Data:** No inconsistencies were found in 'Month' or 'VisitorType' columns, and no unrealistic numerical values were present in duration columns.
- **Outliers:** Outliers in key numerical columns were identified and capped using the IQR method, resulting in a more robust dataset without losing data points.
- **Categorical Encoding:** 'Month' and 'VisitorType' were successfully one-hot encoded, converting them into a numerical format suitable for machine learning models.
- **Feature Scaling:** All relevant numerical features were scaled using `StandardScaler` to ensure they are on a comparable scale, which is crucial for the performance of many machine learning algorithms.

